lib/interferon.rb

```diff
@@ -138,127 +138,143 @@ def update_alerts(destinations, hosts, alerts, groups)
       loader.get_all(destinations).each do |dest|
         break if @request_shutdown
         log.info "updating alerts on #{dest.class.name}"
+        update_alerts_on_destination(dest, hosts, alerts, groups)
+      end
+    end
 
-        # track some counters/stats per destination
-        start_time = Time.new.to_f
-
-        # get already-defined alerts
-        existing_alerts = dest.existing_alerts.dup
-        existing_alerts.each{ |key, existing_alert| existing_alert['still_exists'] = false }
-
-        # create or update alerts; mark when we've done that
-        alerts_queue = Hash.new
-        alerts.each do |alert|
-          break if @request_shutdown
-          counters = {
-            :errors => 0,
-            :evals => 0,
-            :applies => 0,
-            :hosts => hosts.length
-          }
-          last_eval_error = nil
-
-          hosts.each do |hostinfo|
-            begin
-              alert.evaluate(hostinfo)
-              counters[:evals] += 1
-            rescue StandardError => e
-              log.debug "Evaluation of alert #{alert} failed in the context of host #{hostinfo}"
-              counters[:errors] += 1
-              last_eval_error = e
-              next
-            end
+    def update_alerts_on_destination(dest, hosts, alerts, groups)
+      # track some counters/stats per destination
+      start_time = Time.new.to_f
 
-            # don't define an alert that doesn't apply to this hostinfo
-            unless alert[:applies]
-              log.debug "alert #{alert[:name]} doesn't apply to #{hostinfo.inspect}"
-              next
-            end
+      # get already-defined alerts
+      existing_alerts = dest.existing_alerts.dup
+      existing_alerts.each{ |key, existing_alert| existing_alert['still_exists'] = false }
 
-            counters[:applies] += 1
+      alerts_queue = build_alerts_queue(hosts, alerts, groups)
 
-            # don't define alerts twice
-            next if alerts_queue.key?(alert[:name])
+      # flush queue
+      created_alerts_keys = create_alerts(dest, alerts_queue)
+      created_alerts_keys.each do |alert_key|
+        # don't delete alerts we still have defined
+        existing_alerts[alert_key]['still_exists'] = true if existing_alerts.include?(alert_key)
+      end
 
-            # figure out who to notify
-            people = Set.new(alert[:notify][:people])
-            alert[:notify][:groups].each do |g|
-              people += (groups[g] || [])
-            end
+      # remove existing alerts that shouldn't exist
+      to_delete = existing_alerts.reject{ |key, existing_alert| existing_alert['still_exists'] }
```

Comment:
Good point. I think I'd prefer doing this in a different pr in order to minimize the impact of this one. 
The starting point of this pr is that I'll reuse some code here to enable dry run to really hit datadog api's. So I will definitely follow up with a couple of other prs, they'll have this change.